apis/filesystem/src/main/java/org/jclouds/filesystem/strategy/internal/FilesystemStorageStrategyImpl.java

```diff
@@ -964,17 +964,27 @@ private File openFolder(String folderName) throws IOException {
    }
 
    private static void populateBlobKeysInContainer(File directory, Set<String> blobNames,
-         Function<String, String> function) {
+         String prefix, Function<String, String> function) {
       File[] children = directory.listFiles();
       if (children == null) {
          return;
       }
       for (File child : children) {
+         String fullPath = function.apply(child.getAbsolutePath());
          if (child.isFile()) {
-            blobNames.add( function.apply(child.getAbsolutePath()) );
+            if (prefix != null && !fullPath.startsWith(prefix)) {
+               continue;
+            }
+            blobNames.add(fullPath);
          } else if (child.isDirectory()) {
-            blobNames.add(function.apply(child.getAbsolutePath()) + File.separator); // TODO: undo if failures
-            populateBlobKeysInContainer(child, blobNames, function);
+            // Consider a prefix /a/b/c but we have only descended to path /a.
+            // We need to match the path against the prefix to continue
```

Comment:
I was confused by the additional check here, but after removing it and running the tests, it makes sense: the directories will not include the trailing "/", so if the prefix is "dir/", then it wouldn't match the directory "dir" (excluding it from the list and not listing its contents). I think this comment helps, but maybe it could more explicitly call out this specific issues. As far as I can tell, this only affects prefix listings that end with the file separator.

Reply:
Addressed suggestion via a more narrow filter.